metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -195,21 +197,33 @@ void CanonicalDBG::append_next_rc_nodes(node_index node,
 
                 c = kmer::KmerExtractorBOSS::complement(c);
 
-                if (children[c] == npos)
+                if (children[c] != npos) {
+                    if (!k_odd_) {
+                        is_palindrome_cache_.Put(next, true);
+                    } else {
+                        throw std::runtime_error(fmt::format(
+                            "Primary graph contains both forward and reverse complement: {} {} -> {} {}\t{} {}",
+                            node, graph_.get_node_sequence(node),
+                            children[c], graph_.get_node_sequence(children[c]),
+                            next, graph_.get_node_sequence(next)
+                        ));
```

Comment:
why not the conventional `logger->error` + `exit`?

Reply:
I changed it so all errors which derive from bad user input use the logger, whereas I kept the exception throw in one place where it can only be caused by programmer error.

Reply:
This can also be caused by a user's error. You can mark a non-primary graph as primary and then this error will pop up.

Reply:
For errors caused by a programmer's mistake use assertions.
There is no point to have three: exception / error+exit / assert.

Reply:
So I changed this line to be a logger error. The only place where it still throws is in `add_sequence`, where it throws if the input graph is const.

Reply:
anyway, I changed that to an assert as well

Reply:
Ok, we do use exceptions sometimes.

For instance, in that case, it should in principle never happen, so an assert would work, but it's really hard to trace the logic down to parsing commands in CLI, so I would not guarantee that it never happens and throw exceptions instead of writing assertions in such cases.

I hope this makes it clear.